### PR TITLE
Step 3 – generic controlplane: split server

### PR DESCRIPTION
**Pull request by sttts**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

This PR splits the kube-apiserver server struct `Instance` (`pkg/controlplane/instance.go`) into the kube-apiserver-specific and the generic one `pkg/controlplane/apiserver.Server`.

Part of https://github.com/kubernetes/kubernetes/pull/124530.

#### Which issue(s) this PR fixes:

Towards https://github.com/kubernetes/enhancements/issues/4080

#### Special notes for your reviewer:

1. This is exclusively code move. No functional changes.
2. `pkg/controlplane` is a misnamer. It's actually highly kube-apiserver specific and will eventually be renamed to make the difference to `pkg/controlplane/apiserver` clear. But that's beyond this PR.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/4080-generic-controlplane

### CODE REVIEW

**Comment by hailkomputer** [comment]
Hello! Bug triage team lead here. I want to check the status. 
The code freeze is starting [01:00 UTC Wednesday 1st November 2023 / 18:00 PDT Tuesday 31st October 2023](https://everytimezone.com/s/24b27b03) (about two weeks from now), and while there is still plenty of time, we want to ensure that each PR has a chance to be merged and issue’s are addressed on time.
As the PR is targeting 1.29, is it still planned for this release?

**Comment by dims** [comment]
This work-in-progress PR needs a rebase. please rebase if this PR is still needed for the upcoming release.

**Comment by hailkomputer** [comment]
Hello @sttts and @dims, bug triage team lead here again. I want to echo that the code freeze is starting [01:00 UTC Wednesday 1st November 2023 / 18:00 PDT Tuesday 31st October 2023](https://everytimezone.com/s/24b27b03) (next week), and we want to ensure that each PR has a chance to be merged.
Is this PR still targeting the 1.29 release?

**Comment by dims** [comment]
/milestone clear

this won't make it to code freeze @sttts

File: pkg/controlplane/apiserver/apis.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+			// everything else goes to /apis
+			nonLegacy = append(nonLegacy, &apiGroupInfo)
+		}
+	}
+
+	if err := s.GenericAPIServer.InstallAPIGroups(nonLegacy...); err != nil {
+		return fmt.Errorf("error in registering group versions: %w", err)
+	}
+	return nil
+}
```
Comment (sttts) [inline pkg/controlplane/apiserver/apis.go:91]: For easier review:
```patch
 // InstallAPIs will install the APIs for the restStorageProviders if they are enabled.
-func (m *Instance) InstallAPIs(apiResourceConfigSource serverstorage.APIResourceConfigSource, restOptionsGetter generic.RESTOptionsGetter, restStorageProviders ...RESTStorageProvider) error {
+func (s *Server) InstallAPIs(restStorageProviders ...RESTStorageProvider) error {
 	nonLegacy := []*genericapiserver.APIGroupInfo{}

 	// used later in the loop to filter the served resource by those that have expired.
-	resourceExpirationEvaluator, err := genericapiserver.NewResourceExpirationEvaluator(*m.GenericAPIServer.Version)
+	resourceExpirationEvaluator, err := genericapiserver.NewResourceExpirationEvaluator(*s.GenericAPIServer.Version)
 	if err != nil {
 		return err
 	}

 	for _, restStorageBuilder := range restStorageProviders {
 		groupName := restStorageBuilder.GroupName()
-		apiGroupInfo, err := restStorageBuilder.NewRESTStorage(apiResourceConfigSource, restOptionsGetter)
+		apiGroupInfo, err := restStorageBuilder.NewRESTStorage(s.APIResourceConfigSource, s.RESTOptionsGetter)
 		if err != nil {
-			return fmt.Errorf("problem initializing API group %q : %v", groupName, err)
+			return fmt.Errorf("problem initializing API group %q: %w", groupName, err)
 		}
 		if len(apiGroupInfo.VersionedResourcesStorageMap) == 0 {
 			// If we have no storage for any resource configured, this API group is effectively disabled.
@@ -35,14 +35,14 @@
 		if postHookProvider, ok := restStorageBuilder.(genericapiserver.PostStartHookProvider); ok {
 			name, hook, err := postHookProvider.PostStartHook()
 			if err != nil {
-				klog.Fatalf("Error building PostStartHook: %v", err)
+				return fmt.Errorf("error building PostStartHook: %w", err)
 			}
-			m.GenericAPIServer.AddPostStartHookOrDie(name, hook)
+			s.GenericAPIServer.AddPostStartHookOrDie(name, hook)
 		}

 		if len(groupName) == 0 {
 			// the legacy group for core APIs is special that it is installed into /api via this special install method.
-			if err := m.GenericAPIServer.InstallLegacyAPIGroup(genericapiserver.DefaultLegacyAPIPrefix, &apiGroupInfo); err != nil {
+			if err := s.GenericAPIServer.InstallLegacyAPIGroup(genericapiserver.DefaultLegacyAPIPrefix, &apiGroupInfo); err != nil {
 				return fmt.Errorf("error in registering legacy API: %w", err)
 			}
 		} else {
@@ -51,8 +51,8 @@
 		}
 	}

-	if err := m.GenericAPIServer.InstallAPIGroups(nonLegacy...); err != nil {
-		return fmt.Errorf("error in registering group versions: %v", err)
+	if err := s.GenericAPIServer.InstallAPIGroups(nonLegacy...); err != nil {
+		return fmt.Errorf("error in registering group versions: %w", err)
 	}
 	return nil
 }
```

File: pkg/controlplane/apiserver/server.go [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+
+	client := kubernetes.NewForConfigOrDie(s.GenericAPIServer.LoopbackClientConfig)
+	if len(c.SystemNamespaces) > 0 {
+		s.GenericAPIServer.AddPostStartHookOrDie("start-system-namespaces-controller", func(hookContext genericapiserver.PostStartHookContext) error {
+			go systemnamespaces.NewController(c.SystemNamespaces, client, s.VersionedInformers.Core().V1().Namespaces()).Run(hookContext.StopCh)
+			return nil
+		})
+	}
+
+	if utilfeature.DefaultFeatureGate.Enabled(features.UnknownVersionInteroperabilityProxy) {
```
Comment (sttts) [inline pkg/controlplane/apiserver/server.go:143]: For easier review:
```patch
+	_, publicServicePort, err := c.Generic.SecureServing.HostPort()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get listener address: %w", err)
+	}
+
 	if utilfeature.DefaultFeatureGate.Enabled(features.UnknownVersionInteroperabilityProxy) {
-		peeraddress := getPeerAddress(c.ControlPlane.Extra.PeerAdvertiseAddress, c.ControlPlane.Generic.PublicAddress, publicServicePort)
+		peeraddress := getPeerAddress(c.Extra.PeerAdvertiseAddress, c.Generic.PublicAddress, publicServicePort)
 		peerEndpointCtrl := peerreconcilers.New(
-			c.ControlPlane.Generic.APIServerID,
+			c.Generic.APIServerID,
 			peeraddress,
-			c.ControlPlane.Extra.PeerEndpointLeaseReconciler,
-			c.Extra.EndpointReconcilerConfig.Interval,
+			c.Extra.PeerEndpointLeaseReconciler,
+			c.Extra.PeerEndpointReconcileInterval,
 			client)
 		if err != nil {
 			return nil, fmt.Errorf("failed to create peer endpoint lease controller: %w", err)
 		}
-		m.GenericAPIServer.AddPostStartHookOrDie("peer-endpoint-reconciler-controller",
+		s.GenericAPIServer.AddPostStartHookOrDie("peer-endpoint-reconciler-controller",
 			func(hookContext genericapiserver.PostStartHookContext) error {
 				peerEndpointCtrl.Start(hookContext.StopCh)
 				return nil
 			})
-		m.GenericAPIServer.AddPreShutdownHookOrDie("peer-endpoint-reconciler-controller",
+		s.GenericAPIServer.AddPreShutdownHookOrDie("peer-endpoint-reconciler-controller",
 			func() error {
 				peerEndpointCtrl.Stop()
 				return nil
 			})
-		// Add PostStartHooks for Unknown Version Proxy filter.
-		if c.ControlPlane.Extra.PeerProxy != nil {
-			m.GenericAPIServer.AddPostStartHookOrDie("unknown-version-proxy-filter", func(context genericapiserver.PostStartHookContext) error {
-				err := c.ControlPlane.Extra.PeerProxy.WaitForCacheSync(context.StopCh)
+		if c.Extra.PeerProxy != nil {
+			s.GenericAPIServer.AddPostStartHookOrDie("unknown-version-proxy-filter", func(context genericapiserver.PostStartHookContext) error {
+				err := c.Extra.PeerProxy.WaitForCacheSync(context.StopCh)
 				return err
 			})
 		}
 	}

-	m.GenericAPIServer.AddPostStartHookOrDie("start-cluster-authentication-info-controller", func(hookContext genericapiserver.PostStartHookContext) error {
-		controller := clusterauthenticationtrust.NewClusterAuthenticationTrustController(m.ClusterAuthenticationInfo, client)
+	s.GenericAPIServer.AddPostStartHookOrDie("start-cluster-authentication-info-controller", func(hookContext genericapiserver.PostStartHookContext) error {
+		controller := clusterauthenticationtrust.NewClusterAuthenticationTrustController(s.ClusterAuthenticationInfo, client)

 		// generate a context  from stopCh. This is to avoid modifying files which are relying on apiserver
 		// TODO: See if we can pass ctx to the current method
 		ctx := wait.ContextForChannel(hookContext.StopCh)

 		// prime values and start listeners
-		if m.ClusterAuthenticationInfo.ClientCA != nil {
-			m.ClusterAuthenticationInfo.ClientCA.AddListener(controller)
-			if controller, ok := m.ClusterAuthenticationInfo.ClientCA.(dynamiccertificates.ControllerRunner); ok {
+		if s.ClusterAuthenticationInfo.ClientCA != nil {
+			s.ClusterAuthenticationInfo.ClientCA.AddListener(controller)
+			if controller, ok := s.ClusterAuthenticationInfo.ClientCA.(dynamiccertificates.ControllerRunner); ok {
 				// runonce to be sure that we have a value.
 				if err := controller.RunOnce(ctx); err != nil {
 					runtime.HandleError(err)
@@ -46,9 +50,9 @@
 				go controller.Run(ctx, 1)
 			}
 		}
-		if m.ClusterAuthenticationInfo.RequestHeaderCA != nil {
-			m.ClusterAuthenticationInfo.RequestHeaderCA.AddListener(controller)
-			if controller, ok := m.ClusterAuthenticationInfo.RequestHeaderCA.(dynamiccertificates.ControllerRunner); ok {
+		if s.ClusterAuthenticationInfo.RequestHeaderCA != nil {
+			s.ClusterAuthenticationInfo.RequestHeaderCA.AddListener(controller)
+			if controller, ok := s.ClusterAuthenticationInfo.RequestHeaderCA.(dynamiccertificates.ControllerRunner); ok {
 				// runonce to be sure that we have a value.
 				if err := controller.RunOnce(ctx); err != nil {
 					runtime.HandleError(err)
@@ -62,15 +66,15 @@
 	})

 	if utilfeature.DefaultFeatureGate.Enabled(apiserverfeatures.APIServerIdentity) {
-		m.GenericAPIServer.AddPostStartHookOrDie("start-kube-apiserver-identity-lease-controller", func(hookContext genericapiserver.PostStartHookContext) error {
+		s.GenericAPIServer.AddPostStartHookOrDie("start-kube-apiserver-identity-lease-controller", func(hookContext genericapiserver.PostStartHookContext) error {
 			// generate a context  from stopCh. This is to avoid modifying files which are relying on apiserver
 			// TODO: See if we can pass ctx to the current method
 			ctx := wait.ContextForChannel(hookContext.StopCh)

-			leaseName := m.GenericAPIServer.APIServerID
-			holderIdentity := m.GenericAPIServer.APIServerID + "_" + string(uuid.NewUUID())
+			leaseName := s.GenericAPIServer.APIServerID
+			holderIdentity := s.GenericAPIServer.APIServerID + "_" + string(uuid.NewUUID())

-			peeraddress := getPeerAddress(c.ControlPlane.Extra.PeerAdvertiseAddress, c.ControlPlane.Generic.PublicAddress, publicServicePort)
+			peeraddress := getPeerAddress(c.Extra.PeerAdvertiseAddress, c.Generic.PublicAddress, publicServicePort)
 			// must replace ':,[]' in [ip:port] to be able to store this as a valid label value
 			controller := lease.NewController(
 				clock.RealClock{},
@@ -82,28 +86,28 @@
 				leaseName,
 				metav1.NamespaceSystem,
 				// TODO: receive identity label value as a parameter when post start hook is moved to generic apiserver.
-				labelAPIServerHeartbeatFunc(KubeAPIServer, peeraddress))
+				labelAPIServerHeartbeatFunc(name, peeraddress))
 			go controller.Run(ctx)
 			return nil
 		})
 		// TODO: move this into generic apiserver and make the lease identity value configurable
-		m.GenericAPIServer.AddPostStartHookOrDie("start-kube-apiserver-identity-lease-garbage-collector", func(hookContext genericapiserver.PostStartHookContext) error {
+		s.GenericAPIServer.AddPostStartHookOrDie("start-kube-apiserver-identity-lease-garbage-collector", func(hookContext genericapiserver.PostStartHookContext) error {
 			go apiserverleasegc.NewAPIServerLeaseGC(
 				client,
 				IdentityLeaseGCPeriod,
 				metav1.NamespaceSystem,
-				KubeAPIServerIdentityLeaseLabelSelector,
+				IdentityLeaseComponentLabelKey+"="+name,
 			).Run(hookContext.StopCh)
 			return nil
 		})
 	}

-	m.GenericAPIServer.AddPostStartHookOrDie("start-legacy-token-tracking-controller", func(hookContext genericapiserver.PostStartHookContext) error {
+	s.GenericAPIServer.AddPostStartHookOrDie("start-legacy-token-tracking-controller", func(hookContext genericapiserver.PostStartHookContext) error {
 		go legacytokentracking.NewController(client).Run(hookContext.StopCh)
 		return nil
 	})

-	return m, nil
+	return s, nil
 }

 func labelAPIServerHeartbeatFunc(identity string, peeraddress string) lease.ProcessLeaseFunc {
```

File: pkg/controlplane/apiserver/server.go [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+			// expect this feature to be enabled. We log the former case as an Error
+			// and the latter case as an Info.
+			klog.Error(msg)
+		} else {
+			klog.Info(msg)
+		}
+	} else {
+		routes.NewOpenIDMetadataServer(md.ConfigJSON, md.PublicKeysetJSON).
+			Install(generic.Handler.GoRestfulContainer)
+	}
```
Comment (sttts) [inline pkg/controlplane/apiserver/server.go:124]: For easier review:
```patch
-	if c.ControlPlane.Extra.EnableLogsSupport {
-		routes.Logs{}.Install(s.Handler.GoRestfulContainer)
+	if c.EnableLogsSupport {
+		routes.Logs{}.Install(generic.Handler.GoRestfulContainer)
 	}

 	// Metadata and keys are expected to only change across restarts at present,
 	// so we just marshal immediately and serve the cached JSON bytes.
 	md, err := serviceaccount.NewOpenIDMetadata(
-		c.ControlPlane.Extra.ServiceAccountIssuerURL,
-		c.ControlPlane.Extra.ServiceAccountJWKSURI,
-		c.ControlPlane.Generic.ExternalAddress,
-		c.ControlPlane.Extra.ServiceAccountPublicKeys,
+		c.ServiceAccountIssuerURL,
+		c.ServiceAccountJWKSURI,
+		c.Generic.ExternalAddress,
+		c.ServiceAccountPublicKeys,
 	)
 	if err != nil {
 		// If there was an error, skip installing the endpoints and log the
@@ -18,7 +18,7 @@
 		msg := fmt.Sprintf("Could not construct pre-rendered responses for"+
 			" ServiceAccountIssuerDiscovery endpoints. Endpoints will not be"+
 			" enabled. Error: %v", err)
-		if c.ControlPlane.Extra.ServiceAccountIssuerURL != "" {
+		if c.ServiceAccountIssuerURL != "" {
 			// The user likely expects this feature to be enabled if issuer URL is
 			// set and the feature gate is enabled. In the future, if there is no
 			// longer a feature gate and issuer URL is not set, the user may not
@@ -30,5 +30,5 @@
 		}
 	} else {
 		routes.NewOpenIDMetadataServer(md.ConfigJSON, md.PublicKeysetJSON).
-			Install(s.Handler.GoRestfulContainer)
+			Install(generic.Handler.GoRestfulContainer)
 	}
```

File: pkg/controlplane/apiserver/peer.go [hunk 4]
```diff
@@ -87,3 +89,13 @@ func CreatePeerEndpointLeaseReconciler(c genericapiserver.Config, storageFactory
...
+
+// utility function to get the apiserver address that is used by peer apiservers to proxy
+// requests to this apiserver in case the peer is incapable of serving the request
+func getPeerAddress(peerAdvertiseAddress reconcilers.PeerAdvertiseAddress, publicAddress net.IP, publicServicePort int) string {
+	if peerAdvertiseAddress.PeerAdvertiseIP != "" && peerAdvertiseAddress.PeerAdvertisePort != "" {
+		return net.JoinHostPort(peerAdvertiseAddress.PeerAdvertiseIP, peerAdvertiseAddress.PeerAdvertisePort)
+	} else {
+		return net.JoinHostPort(publicAddress.String(), strconv.Itoa(publicServicePort))
+	}
+}
```
Comment (sttts) [inline pkg/controlplane/apiserver/peer.go:101]: For easier review:
```patch
 // utility function to get the apiserver address that is used by peer apiservers to proxy
 // requests to this apiserver in case the peer is incapable of serving the request
-func getPeerAddress(peerAdvertiseAddress peerreconcilers.PeerAdvertiseAddress, publicAddress net.IP, publicServicePort int) string {
+func getPeerAddress(peerAdvertiseAddress reconcilers.PeerAdvertiseAddress, publicAddress net.IP, publicServicePort int) string {
 	if peerAdvertiseAddress.PeerAdvertiseIP != "" && peerAdvertiseAddress.PeerAdvertisePort != "" {
 		return net.JoinHostPort(peerAdvertiseAddress.PeerAdvertiseIP, peerAdvertiseAddress.PeerAdvertisePort)
 	} else {
```

**Comment by mjudeikis** [comment]
looks straightforward move. waiting on https://github.com/kubernetes/kubernetes/pull/124576

**Comment by sttts** [comment]
https://github.com/kubernetes/kubernetes/pull/124576 merged. @mjudeikis if you can take another look

File: test/integration/controlplane/apiserver_identity_test.go [hunk 5]
```diff
@@ -84,7 +85,7 @@ func TestCreateLeaseOnStart(t *testing.T) {
 		leases, err := kubeclient.
 			CoordinationV1().
 			Leases(metav1.NamespaceSystem).
-			List(context.TODO(), metav1.ListOptions{LabelSelector: controlplane.KubeAPIServerIdentityLeaseLabelSelector})
+			List(context.TODO(), metav1.ListOptions{LabelSelector: controlplaneapiserver.IdentityLeaseComponentLabelKey + "=" + controlplane.KubeAPIServer})
```
Comment (mjudeikis) [inline test/integration/controlplane/apiserver_identity_test.go:88]: Just for learning, why this changes?

Comment (sttts) [reply test/integration/controlplane/apiserver_identity_test.go:88]: `KubeAPIServerIdentityLeaseLabelSelector` was a strange constant, although the components were defined just next to it. Also we want other component names than `kube-apiserver` for the generic case.

**Comment by mjudeikis** [comment]
/lgtm 
small changes for the win!

File: pkg/controlplane/apiserver/server.go [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+	"k8s.io/kubernetes/pkg/controlplane/controller/apiserverleasegc"
+	"k8s.io/kubernetes/pkg/controlplane/controller/clusterauthenticationtrust"
+	"k8s.io/kubernetes/pkg/controlplane/controller/legacytokentracking"
+	"k8s.io/kubernetes/pkg/controlplane/controller/systemnamespaces"
+	"k8s.io/kubernetes/pkg/features"
+	"k8s.io/kubernetes/pkg/routes"
+	"k8s.io/kubernetes/pkg/serviceaccount"
+)
+
+var (
```
Comment (luxas) [inline pkg/controlplane/apiserver/server.go:52]: I know this was `var` before as well, but couldn't it be `const`?

File: pkg/controlplane/apiserver/server.go [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+type Server struct {
+	GenericAPIServer *genericapiserver.GenericAPIServer
+
+	APIResourceConfigSource   serverstorage.APIResourceConfigSource
+	RESTOptionsGetter         genericregistry.RESTOptionsGetter
+	ClusterAuthenticationInfo clusterauthenticationtrust.ClusterAuthenticationInfo
+	VersionedInformers        clientgoinformers.SharedInformerFactory
+}
+
+// New returns a new instance of Master from the given config.
```
Comment (luxas) [inline pkg/controlplane/apiserver/server.go:82]: new instance of Server...

File: pkg/controlplane/apiserver/server.go [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+// Certain config fields will be set to a default value if unset.
+// Certain config fields must be specified, including:
+// KubeletClientConfig
```
Comment (luxas) [inline pkg/controlplane/apiserver/server.go:85]: I guess this is not relevant anymore; either we remove this last comment, or specify the fields which are actually required.

File: pkg/controlplane/apiserver/server.go [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+			controller := lease.NewController(
+				clock.RealClock{},
+				client,
+				holderIdentity,
+				int32(IdentityLeaseDurationSeconds),
+				nil,
+				IdentityLeaseRenewIntervalPeriod,
+				leaseName,
+				metav1.NamespaceSystem,
+				// TODO: receive identity label value as a parameter when post start hook is moved to generic apiserver.
```
Comment (luxas) [inline pkg/controlplane/apiserver/server.go:231]: I guess this is fixed now? 😄

File: pkg/controlplane/apiserver/server.go [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+				labelAPIServerHeartbeatFunc(name, peeraddress))
+			go controller.Run(ctx)
+			return nil
+		})
+		// TODO: move this into generic apiserver and make the lease identity value configurable
```
Comment (luxas) [inline pkg/controlplane/apiserver/server.go:236]: do we want to propagate the name variable into the hook ID below, instead of having "kube-apiserver" in the name, or if the IDs do not need to be stable, I guess we can just change "start-kube-apiserver-identity-lease-garbage-collector" to "start-apiserver-identity-lease-garbage-collector"

File: pkg/controlplane/apiserver/server.go [hunk 11]
```diff
@@ -0,0 +1,285 @@
...
+		s.GenericAPIServer.AddPostStartHookOrDie("start-kube-apiserver-identity-lease-garbage-collector", func(hookContext genericapiserver.PostStartHookContext) error {
+			go apiserverleasegc.NewAPIServerLeaseGC(
+				client,
+				IdentityLeaseGCPeriod,
+				metav1.NamespaceSystem,
```
Comment (luxas) [inline pkg/controlplane/apiserver/server.go:241]: shall this become configurable in the future, like how SystemNamespaces are configurable already?

File: pkg/controlplane/apiserver/server.go [hunk 12]
```diff
@@ -0,0 +1,285 @@
...
+			if controller, ok := s.ClusterAuthenticationInfo.RequestHeaderCA.(dynamiccertificates.ControllerRunner); ok {
+				// runonce to be sure that we have a value.
+				if err := controller.RunOnce(ctx); err != nil {
+					runtime.HandleError(err)
+				}
+				go controller.Run(ctx, 1)
+			}
+		}
+
+		go controller.Run(ctx, 1)
```
Comment (luxas) [inline pkg/controlplane/apiserver/server.go:207]: maybe it's just me, but it seems scary to have the same variable name here as for the "inner" controllers, that define "local" variables with the same name

I know the code was like this, but why not just (later) move this line up where controller was defined, and use non-overlapping names?

File: pkg/controlplane/apiserver/server.go [hunk 13]
```diff
@@ -0,0 +1,285 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("failed to get listener address: %w", err)
+	}
+
+	if utilfeature.DefaultFeatureGate.Enabled(features.UnknownVersionInteroperabilityProxy) {
+		peeraddress := getPeerAddress(c.Extra.PeerAdvertiseAddress, c.Generic.PublicAddress, publicServicePort)
+		peerEndpointCtrl := peerreconcilers.New(
+			c.Generic.APIServerID,
+			peeraddress,
+			c.Extra.PeerEndpointLeaseReconciler,
```
Comment (luxas) [inline pkg/controlplane/apiserver/server.go:154]: I guess you could remove the `Extra` field reference now if you did not want that to be visible in the field path

File: pkg/controlplane/apiserver/server.go [hunk 14]
```diff
@@ -0,0 +1,285 @@
...
+				client,
+				IdentityLeaseGCPeriod,
+				metav1.NamespaceSystem,
+				IdentityLeaseComponentLabelKey+"="+name,
+			).Run(hookContext.StopCh)
+			return nil
+		})
+	}
+
+	s.GenericAPIServer.AddPostStartHookOrDie("start-legacy-token-tracking-controller", func(hookContext genericapiserver.PostStartHookContext) error {
```
Comment (luxas) [inline pkg/controlplane/apiserver/server.go:248]: why do we need this in a new generic API server, if it's "legacy"?

File: pkg/controlplane/apiserver/apis.go [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+func (s *Server) InstallAPIs(restStorageProviders ...RESTStorageProvider) error {
+	nonLegacy := []*genericapiserver.APIGroupInfo{}
+
+	// used later in the loop to filter the served resource by those that have expired.
+	resourceExpirationEvaluator, err := genericapiserver.NewResourceExpirationEvaluator(*s.GenericAPIServer.Version)
+	if err != nil {
+		return err
+	}
+
+	for _, restStorageBuilder := range restStorageProviders {
```
Comment (luxas) [inline pkg/controlplane/apiserver/apis.go:44]: nit: (probably was before as well) unexpected to call it `restStorageBuilder` when it is in the `restStorageProviders` list, `restStorageProvider` would be more expected

File: pkg/controlplane/apiserver/server.go [hunk 16]
```diff
@@ -0,0 +1,285 @@
...
+		APIResourceConfigSource:   c.APIResourceConfigSource,
+		RESTOptionsGetter:         c.Generic.RESTOptionsGetter,
+		ClusterAuthenticationInfo: c.ClusterAuthenticationInfo,
+		VersionedInformers:        c.VersionedInformers,
+	}
+
+	client := kubernetes.NewForConfigOrDie(s.GenericAPIServer.LoopbackClientConfig)
+	if len(c.SystemNamespaces) > 0 {
+		s.GenericAPIServer.AddPostStartHookOrDie("start-system-namespaces-controller", func(hookContext genericapiserver.PostStartHookContext) error {
+			go systemnamespaces.NewController(c.SystemNamespaces, client, s.VersionedInformers.Core().V1().Namespaces()).Run(hookContext.StopCh)
```
Comment (luxas) [inline pkg/controlplane/apiserver/server.go:139]: probably this is the pattern we want to follow, to have "local" properties off commonly-used fields in the structs they are used, but just noting that we could just use `c.VersionedInformers` here, because it's not used anywhere else (for now at least)

similar case for ClusterAuthenticationInfo, it's only used in this setup function, not later, so technically it would not need to be in the struct

**Review by luxas — COMMENTED** [review]
Thanks @sttts!

Minor comments, mostly for the future
Indeed LGTM as well 👍
